### PR TITLE
Hexen's Audit Addressal + Further testing

### PR DESCRIPTION
**Pull request by Sidu28**
Associated ticket: https://github.com/Layr-Labs/eignlayr-contracts/issues/981

### CODE REVIEW

File: src/contracts/pods/EigenPod.sol [hunk 1]
```diff
@@ -667,20 +669,19 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
          * in the beacon chain as a full withdrawal.  Thus such a validator can prove another full withdrawal, and
          * withdraw that ETH via the queuedWithdrawal flow in the strategy manager.
          */
-            // if the withdrawal amount is greater than the MAX_VALIDATOR_BALANCE_GWEI (i.e. the max amount restaked on EigenLayer, per ETH validator)
-        uint64 maxRestakedBalanceGwei = _calculateRestakedBalanceGwei(MAX_VALIDATOR_BALANCE_GWEI);
```
Comment (Sidu28) [inline src/contracts/pods/EigenPod.sol:671]: addressing the bug I found after the hexens report, where we send up to 1 ETH extra instantly to the podOwner vs sending it via the queuedWithdrawal system

File: src/contracts/libraries/BeaconChainProofs.sol [hunk 2]
```diff
@@ -285,6 +285,11 @@ library BeaconChainProofs {
             "BeaconChainProofs.verifyWithdrawal: withdrawalIndex is too large"
         );
 
+        require(
+            withdrawalProof.historicalSummaryIndex < 2 ** HISTORICAL_SUMMARIES_TREE_HEIGHT,
+            "BeaconChainProofs.verifyWithdrawal: historicalSummaryIndex is too large"
+        );
+
```
Comment (Sidu28) [inline src/contracts/libraries/BeaconChainProofs.sol:292]: This is the other hexens finding, EIG 10.

File: src/contracts/pods/EigenPod.sol [hunk 3]
```diff
@@ -667,20 +669,18 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
...
-            // and the extra execution layer ETH in the contract is MAX_VALIDATOR_BALANCE_GWEI, which must be withdrawn through EigenLayer's normal withdrawal process
-            withdrawableRestakedExecutionLayerGwei += maxRestakedBalanceGwei;
-            withdrawalAmountWei = maxRestakedBalanceGwei * GWEI_TO_WEI;
+            // and the extra execution layer ETH in the contract is MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR, which must be withdrawn through EigenLayer's normal withdrawal process
+            withdrawableRestakedExecutionLayerGwei += MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR;
+            withdrawalAmountWei = MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR * GWEI_TO_WEI;
         } else {
             // otherwise, just use the full withdrawal amount to continue to "back" the podOwner's remaining shares in EigenLayer
             // (i.e. none is instantly withdrawable)
-            withdrawalAmountGwei = _calculateRestakedBalanceGwei(withdrawalAmountGwei);
```
Comment (Sidu28) [inline src/contracts/pods/EigenPod.sol:683]: This change addresses EIG 7

File: docs/core/EigenPodManager.md [hunk 4]
```diff
@@ -59,7 +59,7 @@ Note: the functions of the `EigenPodManager` and `EigenPod` contracts are tightl
         * The calculation subtracts an offset (`RESTAKED_BALANCE_OFFSET_GWEI`) from the validator's proven balance, and round down to the nearest ETH
         * Related: `uint64 RESTAKED_BALANCE_OFFSET_GWEI`
             * As of M2, this is 0.75 ETH (in Gwei)
-        * Related: `uint64 MAX_VALIDATOR_BALANCE_GWEI`
+        * Related: `uint64 MAX_RESTAKED_BALANCE_GWEI`
```
Comment (wadealexc) [inline docs/core/EigenPodManager.md:62]: I believe this is called `MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR`, per the changes below

Comment (Sidu28) [reply docs/core/EigenPodManager.md]: fixed

File: src/test/EigenPod.t.sol [hunk 5]
```diff
@@ -237,7 +245,9 @@ contract EigenPodTests is ProofParsing, EigenPodPausingConstants {
              strategyManager
         );
 
-        cheats.deal(address(podOwner), 5*stakeAmount);     
+        cheats.deal(address(podOwner), 5*stakeAmount);   
+
+        emit log("WHT is happ");
```
Comment (ChaoticWalrus) [inline src/test/EigenPod.t.sol:250]: looks like there's a few logs you forgot to delete? 😄

File: src/test/EigenPod.t.sol [hunk 6]
```diff
@@ -276,18 +287,10 @@ contract EigenPodTests is ProofParsing, EigenPodPausingConstants {
         require(pod.mostRecentWithdrawalTimestamp() == uint64(block.timestamp), "Most recent withdrawal block number not updated");
     }
 
-
-    function testCheckThatHasRestakedIsSetToTrue() public {
-        testStaking();
-        IEigenPod pod = eigenPodManager.getPod(podOwner);
-        require(pod.hasRestaked() == true, "Pod should not be restaked");
-    }
```
Comment (ChaoticWalrus) [inline src/test/EigenPod.t.sol:284]: we probably still want this test for new pods

File: src/test/EigenPod.t.sol [hunk 7]
```diff
@@ -371,15 +362,6 @@ contract EigenPodTests is ProofParsing, EigenPodPausingConstants {
...
     }
 
-    function testAttemptedWithdrawalAfterVerifyingWithdrawalCredentials() public {
-        testDeployAndVerifyNewEigenPod();
-        IEigenPod pod = eigenPodManager.getPod(podOwner);
-        cheats.startPrank(podOwner);
-        cheats.expectRevert(bytes("EigenPod.hasNeverRestaked: restaking is enabled"));
-        IEigenPod(pod).withdrawBeforeRestaking();
-        cheats.stopPrank();
-    }
```
Comment (ChaoticWalrus) [inline src/test/EigenPod.t.sol:381]: perhaps the test here should be:

1. confirm 'hasRestaked = false'
2. prove withdrawal credentials
3. check that 'hasRestaked = false', as well as any other effects (like nonBeaconChainETH balance being sent and zeroed out)

File: src/contracts/libraries/BeaconChainProofs.sol (hunk 2 above)
Comment (ChaoticWalrus) [reply src/contracts/libraries/BeaconChainProofs.sol:292]: gotcha, so the issue here was being able to select an invalid `historicalSummaryIndex` ? Looks like test coverage was added too 👍

File: src/test/EigenPod.t.sol (hunk 6 above)
Comment (ChaoticWalrus) [reply src/test/EigenPod.t.sol:284]: ah, it got moved into the unit tests 👍

File: src/contracts/pods/EigenPod.sol (hunk 3 above)
Comment (ChaoticWalrus) [reply src/contracts/pods/EigenPod.sol:683]: Is there a specific regression test for this? If so, which test(s) is it?

**Review by ChaoticWalrus — CHANGES_REQUESTED** [review]
Looks quite good.
Plan is apparently to merge a couple other PRs first, so this is still marked as a 'draft' (if I understand correctly?).
Added a couple questions / comments

File: src/test/EigenPod.t.sol (hunk 5 above)
Comment (Sidu28) [reply src/test/EigenPod.t.sol:246]: lol removed

File: src/test/EigenPod.t.sol [hunk 8]
```diff
@@ -338,18 +341,6 @@ contract EigenPodTests is ProofParsing, EigenPodPausingConstants {
...
-        // ./solidityProofGen "ValidatorFieldsProof" 302913 true "data/withdrawal_proof_goerli/goerli_slot_6399999.json"  "data/withdrawal_proof_goerli/goerli_slot_6399998.json" "withdrawal_credential_proof_510257.json"
-         setJSON("./src/test/test-data/withdrawal_credential_proof_302913.json");
-
-        IEigenPod pod = _testDeployAndVerifyNewEigenPod(podOwner, signature, depositDataRoot);
-
-        cheats.expectRevert(bytes("EigenPod.hasNeverRestaked: restaking is enabled"));
-        cheats.startPrank(podOwner);
-        pod.withdrawBeforeRestaking();
-        cheats.stopPrank();
-    }
```
[the remark this answers is not in the document]
Comment (Sidu28) [reply src/test/EigenPod.t.sol:351]: not deleted, moved to new file

File: src/contracts/pods/EigenPod.sol (hunk 1 above)
Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol:671]: this check: https://github.com/Layr-Labs/eigenlayer-contracts/blob/9af7c3a4be085449ee6b4f54b89d5e0611314b85/src/test/EigenPod.t.sol#L1218

File: src/contracts/pods/EigenPod.sol (hunk 3 above)
Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol:683]: https://github.com/Layr-Labs/eigenlayer-contracts/blob/873362f84da939e355f1d9d552b7e6c829198b40/src/test/unit/EigenPodUnit.t.sol#L81C10-L81C10

File: src/test/EigenPod.t.sol (hunk 7 above)
Comment (Sidu28) [reply src/test/EigenPod.t.sol:381]: I dont think I follow - this is just trying to call funcstion that shouldn't be called once you've verified your pod WC

Comment (ChaoticWalrus) [reply src/test/EigenPod.t.sol:381]: mm OK. guess the suggested checks should go in `testDeployAndVerifyNewEigenPod` or similar then instead.

**Review by ChaoticWalrus — APPROVED** [review]

**Pull request merged**